lib/ipmi_backend_utils.pm

```diff
@@ -390,4 +395,84 @@ sub ipmitool {
+...
+mem_encrypt=on turns on the SME memory encryption feature on the host which protects 
+against the physical attack on the hypervisor memory. The kvm_amd.sev parameter 
+actually enables SEV in the kvm module. This subroutine receives only two arguments,
+the dst_machine is the host on which operations will be performed, the root_dir is 
+the partition on which grub files reside. If these two arguments are not given any
+values, operations will be performed on localhost and '/' partition. This subroutine
+calls add_kernel_options to do the actual kernel options adding work.
+
+=cut
+sub enable_sev_in_kernel {
+    my (%args) = @_;
+
+    $args{dst_machine} //= 'localhost';
+    $args{root_dir} //= '';
+    $args{root_dir} .= '/' unless $args{root_dir} =~ /\/$/;
+    croak("No AMD EPYC cpu on $args{dst_machine}, so sev can not be enabled in kernel.") unless (script_run("lscpu | grep -i \'AMD EPYC\'") == 0);
+    add_kernel_options(dst_machine => $args{dst_machine}, root_dir => $args{root_dir}, kernel_opts => 'mem_encrypt=on kvm_amd.sev=1');
+}
+
+=head2 add_kernel_options
+
+  add_kernel_options(dst_machine => $dst_machine, root_dir => $root_dir, kernel_opts => $options)
+
+Adding additional kernel options onto kernel command line in grub config file and 
+also GRUB_CMDLINE_LINUX_DEFAULT line in default grub config file. This subroutine 
+receives only three arguments, dst_machine is the host on which operations will be 
+performed, root_dir is the partition on which grub files reside, kernel_opts holds 
+a single text string this is composed of kernel options separated by spaces. If no
+values are passed in to dst_machine and root_dir, operations will be performed on 
+localhost and '/' partition by default.  Almost all regular kernel options can be
+passed in directly without modification except for very extreme cases in which very
+special characters should be treated specially and even escaped before being used.
+
+=cut
+sub add_kernel_options {
+    my (%args) = @_;
+
+    $args{dst_machine} //= 'localhost';
+    $args{root_dir} //= '';
+    $args{root_dir} .= '/' unless $args{root_dir} =~ /\/$/;
+    $args{kernel_opts} //= '';
+    croak("Nothing to added onto kernel command line. Argument kernel_opts should not be empty.") if ($args{kernel_opts} eq '');
+
+    my @options = split(/ /, $args{kernel_opts});
+    my $cmd = '';
+    my $grub_cfg_file = "$args{root_dir}boot/grub2/grub.cfg";
+    foreach (@options) {
+        $cmd = "sed -i -r \'s/$_//g; /^\\s{1,}linux\\s{1,}\\\/boot\\\// s/\$/ $_/g\' $grub_cfg_file";
+        $cmd = "ssh root\@$args{dst_machine} " . "\"$cmd\"" if ($args{dst_machine} ne 'localhost');
+        assert_script_run($cmd);
+        save_screenshot;
+    }
+    $cmd = "cat $grub_cfg_file";
+    $cmd = "ssh root\@$args{dst_machine} " . "\"$cmd\"" if ($args{dst_machine} ne 'localhost');
+    record_info("Content of $grub_cfg_file on $args{dst_machine}", script_output($cmd, proceed_on_failure => 1));
```

Comment:
Can you make this grub.cfg edit optional? This will make the function more of a general purpose, which can fully follow document guide well(recommend to edit /etc/default/grub), and meanwhile take care of our current bad serial console edit?

Reply:
add_kernel_options already modifies /etc/default/grub. Make what optional ?  Option to do not edit 
```
my $grub_cfg_file = "$args{root_dir}boot/grub2/grub.cfg";
```
?

Reply:
As the function name suggests "add_kernel_options" which handles kernel options on kernel command line.

Reply:
> add_kernel_options already modifies /etc/default/grub. Make what optional ? Option to do not edit
> 
> ```
> my $grub_cfg_file = "$args{root_dir}boot/grub2/grub.cfg";
> ```
> 
> ?

Sorry for the confusion. I mean the block from line 453 to line 464, which  make changes on grub.cfg, we'd better make it optional, eg add a function parameter as switch, if the function is called with that parameter set to 1, it edits both grub.cfg and etc/default/grub, otherwise skip grub.cfg  edit and change etc/default/grub only(the recommended way by suse doc).  After we correct serial console grub edit to the recommended way, we will need to call the function with that switch off.

Reply:
OK. I can add such switch. For exmaple, "edit_grub_config => 1/0" or similar.

Reply:
Done.